### PR TITLE
Extract parsing in data-prepper-core to data-prepper-pipeline-parser module

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
This change extracts the classes for parsing a pipeline configuration and validating the general configuration structure. The main classes extracted include `PipelineConfigurationValidator` and `PipelinesDataFlowModelParser`
 
### Issues Resolved
Related to #2573 
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-pipeline-parser/build.gradle [hunk 1]
```diff
@@ -0,0 +1,9 @@
+dependencies {
+    implementation project(':data-prepper-api')
+    implementation project(':data-prepper-plugins:blocking-buffer')
+    testImplementation platform('org.junit:junit-bom:5.9.1')
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/build.gradle:4]: You don't need this line. The bom is provided in the root project.

File: data-prepper-pipeline-parser/build.gradle [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+    implementation 'com.fasterxml.jackson.core:jackson-databind'
+    implementation 'com.fasterxml.jackson.dataformat:jackson-dataformat-yaml'
+    implementation 'org.apache.commons:commons-collections4:4.4'
+    testImplementation 'org.junit.jupiter:junit-jupiter'
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/build.gradle:8]: I don't think you need this line either.

File: data-prepper-pipeline-parser/src/test/java/org/opensearch/dataprepper/pipeline/parser/TestConfigurationProvider.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class TestConfigurationProvider {
```
Comment (dlvenable) [inline data-prepper-pipeline-parser/src/test/java/org/opensearch/dataprepper/pipeline/parser/TestConfigurationProvider.java:23]: Is this file copied and pasted? If so, can we consolidate it somehow?

**Review by dlvenable — CHANGES_REQUESTED** [review]
I see a number of newly added files that seem that they may be copied and pasted. Can we consolidate?

**Review by dlvenable — APPROVED** [review]
Thanks for your work on this!

**Review by kkondaka — APPROVED** [review]

**Pull request merged**